protocol/daemons/server/types/health_monitor.go

```diff
@@ -0,0 +1,223 @@
+			// where the monitor is stopped before all daemon services have been registered.
+			if stoppable, ok := checkable.(Stoppable); ok {
+				stoppable.Stop()
+			}
+		}
+	}()
+
+	// Enter into the critical section.
+	ms.Lock()
+	defer ms.Unlock()
+
+	// Don't register daemon services if the monitor has been disabled.
+	if ms.disabled {
+		return nil
+	}
+
+	// Don't register additional daemon services if the monitor has already been stopped.
+	// This could be a concern for short-running integration test cases, where the network
+	// stops before all daemon services have been registered.
+	if ms.stopped {
+		// Toggle the stopService flag to true so that the service is stopped after the lock is released.
+		stopService = true
+		return nil
+	}
+
+	if _, ok := ms.serviceToHealthChecker[checkable.ServiceName()]; ok {
+		return fmt.Errorf("service %v already registered", checkable.ServiceName())
+	}
+
+	ms.serviceToHealthChecker[checkable.ServiceName()] = lazyHealthCheckerCreator()
+	return nil
+}
+
+// HealthMonitor monitors the health of daemon services, which implement the HealthCheckable interface. If a
+// registered health-checkable service sustains an unhealthy state for the maximum acceptable unhealthy duration,
+// the monitor will execute a callback function.
+type HealthMonitor struct {
+	mutableState *healthMonitorMutableState
+
+	// These fields are initialized in NewHealthMonitor and are not modified after initialization.
+	logger             log.Logger
+	startupGracePeriod time.Duration
+	pollingFrequency   time.Duration
+}
+
+// NewHealthMonitor creates a new health monitor.
+func NewHealthMonitor(
+	startupGracePeriod time.Duration,
+	pollingFrequency time.Duration,
+	logger log.Logger,
+) *HealthMonitor {
+	return &HealthMonitor{
+		mutableState:       newHealthMonitorMutableState(),
+		logger:             logger.With(cosmoslog.ModuleKey, HealthMonitorLogModuleName),
+		startupGracePeriod: startupGracePeriod,
+		pollingFrequency:   pollingFrequency,
+	}
+}
+
+func (hm *HealthMonitor) DisableForTesting() {
+	hm.mutableState.DisableForTesting()
+}
+
+// RegisterServiceWithCallback registers a HealthCheckable with the health monitor. If the service
+// stays unhealthy every time it is polled during the maximum acceptable unhealthy duration, the monitor will
+// execute the callback function.
+// This method is synchronized. The method returns an error if the service was already registered or the
+// monitor has already been stopped. If the monitor has been stopped, this method will proactively stop the
+// health-checkable service before returning.
+func (hm *HealthMonitor) RegisterServiceWithCallback(
+	hc types.HealthCheckable,
+	maximumAcceptableUnhealthyDuration time.Duration,
+	callback func(error),
+) error {
+	if maximumAcceptableUnhealthyDuration <= 0 {
+		return fmt.Errorf(
+			"health check registration failure for service %v: "+
+				"maximum acceptable unhealthy duration %v must be positive",
```

Comment:
nit: just to confirm, will you be updating this text too to match `maxDaemonUnhealthyDuration`?

Reply:
yes, it's updated [here](https://github.com/dydxprotocol/v4-chain/pull/802/files#diff-9dca63687afbddbc7f90992fa632c143333f1f33d9604b4af8445ee5a5d36342)